syntax = "proto3";

option java_package = "ca.eandb.jmist.proto";
option java_outer_classname = "LightProtos";

import "core.proto";

message PointLight {
  Vector3 position = 1;
}

message AreaLight {
  Vector3 center = 1;
  Vector3 u = 2;
  Vector3 v = 3;
  double size_u = 4;
  double size_v = 5;
}

message DirectionalLight {
  Vector3 direction = 1;
}

message Light {
  enum Type {
    POINT = 0;
    AREA = 1;
    DIRECTIONAL = 2;
  }
  Type type = 1;

  double energy = 2;
  Color color = 3;
  bool shadow = 4;

  oneof parameters {
    PointLight point_light = 5;
    AreaLight area_light = 6;
    DirectionalLight directional_light = 7;
  }
}
